File: home/src/spl/home/mux.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (c) 2026 sol pbc

"""Multiplex driver: framing-layer state + per-stream asyncio I/O.

Bytes in from the TLS-plaintext side are fed into this module; we produce
frames to send back, and each logical stream surfaces as an
`asyncio.StreamReader`/`StreamWriter` pair that the HTTP app can drive.

Flow-control uses the 1 MiB initial window per ../../proto/framing.md §flow
control — this side grants credit as bytes drain into the app; the peer
uses its granted credit to send more data. For MVP the default "grant on
every drained chunk" policy is fine; alpha may revisit.

Concurrent stream cap: 256 per direction. OPENs beyond cap RESET with
STREAM_LIMIT_EXCEEDED (same applies if we originate, but the home rarely
opens streams in v1).
"""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from .framing import (
    FLAG_CLOSE,
    FLAG_DATA,
    FLAG_OPEN,
    FLAG_RESET,
    FLAG_WINDOW,
    INITIAL_WINDOW,
    MAX_CONCURRENT_STREAMS,
    RECOMMENDED_CHUNK,
    RESET_FLOW_CONTROL_ERROR,
    RESET_INTERNAL_ERROR,
    RESET_PROTOCOL_ERROR,
    RESET_STREAM_LIMIT_EXCEEDED,
    Frame,
    FrameDecoder,
    ProtocolError,
    build_close,
    build_data,
    build_open,
    build_reset,
    build_window,
    parse_reset_reason,
    parse_window_credit,
)

if TYPE_CHECKING:
    StreamHandler = Callable[[asyncio.StreamReader, "StreamWriter"], Awaitable[None]]
else:
    StreamHandler = object  # runtime placeholder; only used for type hints


@dataclass
class _StreamState:
    stream_id: int
    reader: asyncio.StreamReader
    reader_closed: bool = False
    writer_closed: bool = False
    # Remaining credit we have to send TO the peer on this stream.
    send_credit: int = INITIAL_WINDOW
    # Credit the peer has to send TO us on this stream.
    recv_credit: int = INITIAL_WINDOW
    # Bytes received since last window update — used to batch WINDOW grants.
    unacked_recv: int = 0
    # Pending send queue: segments we'd emit when credit permits. Kept small;
    # the writer yields until credit is available via an asyncio.Event.
    credit_event: asyncio.Event = field(default_factory=asyncio.Event)
    task: asyncio.Task[None] | None = None


class StreamWriter:
    """Per-stream writer. Calls into the mux to emit DATA/CLOSE/RESET frames."""

    def __init__(self, mux: Multiplexer, state: _StreamState) -> None:
        self._mux = mux
        self._state = state

    async def write(self, data: bytes) -> None:
        if self._state.writer_closed:
            raise ConnectionError(f"stream {self._state.stream_id} writer is closed")
        view = memoryview(data)
        while view:
            chunk_len = min(len(view), RECOMMENDED_CHUNK, self._state.send_credit)
            if chunk_len <= 0:
                self._state.credit_event.clear()
                await self._state.credit_event.wait()
                continue
            chunk = bytes(view[:chunk_len])
            view = view[chunk_len:]
            self._state.send_credit -= chunk_len
            await self._mux._emit(build_data(self._state.stream_id, chunk))

    async def close(self) -> None:
        if self._state.writer_closed:
            return
        self._state.writer_closed = True
        await self._mux._emit(build_close(self._state.stream_id))

    async def reset(self, reason: int = RESET_INTERNAL_ERROR) -> None:
        if self._state.writer_closed and self._state.reader_closed:
            return
        self._state.writer_closed = True
        self._state.reader_closed = True
        await self._mux._emit(build_reset(self._state.stream_id, reason))
        self._state.reader.feed_eof()
        self._mux._forget(self._state.stream_id)


class Multiplexer:
    """Frame-level state. Caller pumps incoming bytes with `feed_ciphertext`."""

    def __init__(
        self,
        send_frame: Callable[[bytes], Awaitable[None]],
        handler: StreamHandler,
        *,
        is_listener: bool = True,
    ) -> None:
        """If `is_listener=True`, this side expects odd stream_ids from the peer."""
        self._decoder = FrameDecoder()
        self._send_frame = send_frame
        self._handler = handler
        self._is_listener = is_listener
        self._streams: dict[int, _StreamState] = {}
        self._closed = False

    async def feed(self, plaintext: bytes) -> None:
        """Feed plaintext bytes from the TLS layer into the framing decoder."""
        if not plaintext:
            return
        self._decoder.feed(plaintext)
        while True:
            try:
                frame = self._decoder.next()
            except ProtocolError:
                # Malformed at the framing layer — abort the whole mux.
                await self._reset_all(RESET_PROTOCOL_ERROR)
                return
            if frame is None:
                return
            await self._dispatch(frame)

    async def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        for state in list(self._streams.values()):
            state.reader.feed_eof()
            state.writer_closed = True
            if state.task and not state.task.done():
                state.task.cancel()
        self._streams.clear()

    # --- internals ---

    async def _dispatch(self, frame: Frame) -> None:
        if frame.flags & FLAG_OPEN:
            if not self._valid_peer_stream_id(frame.stream_id):
                await self._emit(build_reset(frame.stream_id, RESET_PROTOCOL_ERROR))
                return
            if frame.stream_id in self._streams:
                await self._emit(build_reset(frame.stream_id, RESET_PROTOCOL_ERROR))
                return
            if len(self._streams) >= MAX_CONCURRENT_STREAMS:
                await self._emit(build_reset(frame.stream_id, RESET_STREAM_LIMIT_EXCEEDED))
                return
            state = self._open_stream(frame.stream_id)
            if frame.payload:
                state.reader.feed_data(frame.payload)
                state.recv_credit -= len(frame.payload)
            if frame.flags & FLAG_CLOSE:
                state.reader.feed_eof()
                state.reader_closed = True
            return

        maybe_state = self._streams.get(frame.stream_id)
        if maybe_state is None:
            # Spec: unknown stream → RESET with PROTOCOL_ERROR.
            await self._emit(build_reset(frame.stream_id, RESET_PROTOCOL_ERROR))
            return
        state = maybe_state

        if frame.flags & FLAG_DATA:
            if len(frame.payload) > state.recv_credit:
                await self._emit(build_reset(frame.stream_id, RESET_FLOW_CONTROL_ERROR))
                self._terminate(state)
                return
            state.reader.feed_data(frame.payload)
            state.recv_credit -= len(frame.payload)
            state.unacked_recv += len(frame.payload)
            # Grant credit back as we receive — generous policy: restore the
            # full window once we've received half of it.
            if state.unacked_recv >= INITIAL_WINDOW // 2:
                grant = state.unacked_recv
                state.recv_credit += grant
                state.unacked_recv = 0
                await self._emit(build_window(frame.stream_id, grant))
        if frame.flags & FLAG_CLOSE:
            state.reader.feed_eof()
            state.reader_closed = True
            if state.writer_closed:
                self._forget(frame.stream_id)
        if frame.flags & FLAG_WINDOW:
            try:
                credit = parse_window_credit(frame)
            except ProtocolError:
                await self._emit(build_reset(frame.stream_id, RESET_PROTOCOL_ERROR))
                self._terminate(state)
                return
            state.send_credit += credit
            state.credit_event.set()
        if frame.flags & FLAG_RESET:
            try:
                _ = parse_reset_reason(frame)
            except ProtocolError:
                pass  # Unknown reasons tolerated per spec.
            state.reader.feed_eof()
            self._terminate(state)

    def _open_stream(self, stream_id: int) -> _StreamState:
        reader = asyncio.StreamReader()
        state = _StreamState(stream_id=stream_id, reader=reader)
        state.credit_event.set()
        self._streams[stream_id] = state
        writer = StreamWriter(self, state)

        async def runner() -> None:
            try:
                await self._handler(reader, writer)
            except Exception:  # noqa: BLE001 — per-stream errors must not kill the mux
                await writer.reset(RESET_INTERNAL_ERROR)
            finally:
                if not state.writer_closed:
                    try:
                        await writer.close()
                    except Exception:  # noqa: BLE001,S110 — best-effort cleanup
                        pass
                self._forget(stream_id)

        state.task = asyncio.create_task(runner(), name=f"spl-stream-{stream_id}")
        return state

    def _terminate(self, state: _StreamState) -> None:
        state.writer_closed = True
        state.reader_closed = True
        if state.task and not state.task.done():
            state.task.cancel()
        self._forget(state.stream_id)

    def _forget(self, stream_id: int) -> None:
        self._streams.pop(stream_id, None)

    def _valid_peer_stream_id(self, stream_id: int) -> bool:
        # Per proto/framing.md §id allocation: dialing side uses odd ids,
        # listening side uses even. stream_id 0 is reserved for control.
        if stream_id == 0:
            return False
        return (stream_id % 2 == 1) if self._is_listener else (stream_id % 2 == 0)

    async def _emit(self, frame: Frame) -> None:
        if self._closed:
            return
        try:
            encoded = frame.encode()
        except ProtocolError:
            return
        await self._send_frame(encoded)

    async def _reset_all(self, reason: int) -> None:
        for state in list(self._streams.values()):
            await self._emit(build_reset(state.stream_id, reason))
            self._terminate(state)

    # For tests / tunnel-origination: open a new stream from this side.
    async def open_stream(
        self,
        initial_payload: bytes = b"",
    ) -> tuple[asyncio.StreamReader, StreamWriter]:
        next_id = self._next_local_stream_id()
        reader = asyncio.StreamReader()
        state = _StreamState(stream_id=next_id, reader=reader)
        state.credit_event.set()
        self._streams[next_id] = state
        writer = StreamWriter(self, state)
        frame = build_open(next_id, initial_payload)
        if initial_payload:
            state.send_credit -= len(initial_payload)
        await self._emit(frame)
        return reader, writer

    def _next_local_stream_id(self) -> int:
        # listener → even; dialer → odd. Start at 2 or 1 respectively; skip
        # ids already in the map.
        start = 2 if self._is_listener else 1
        cur = start
        while cur in self._streams:
            cur += 2
            if cur > 0xFFFFFFFF:
                raise RuntimeError("stream_id space exhausted")
        return cur

File: mobile/src/mux.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (c) 2026 sol pbc

// Multiplex driver — dialer side.
//
// On the dialing side (mobile), we originate odd-numbered stream IDs per
// proto/framing.md §"id allocation". Each stream is a bidirectional
// byte pipe the caller drives via `openStream()` → `{ write, read, close }`.
//
// Kept deliberately spartan: enough to open 10+ concurrent streams, send
// HTTP requests, collect responses, and exercise credit-based flow
// control. The home-side mux is richer (handler dispatch, protocol-error
// RESET on unexpected OPENs, etc.) because it has to accept arbitrary
// peer behavior; the dialer only originates streams we asked for.

import {
	FLAG_CLOSE,
	FLAG_DATA,
	FLAG_OPEN,
	FLAG_RESET,
	FLAG_WINDOW,
	FrameDecoder,
	INITIAL_WINDOW,
	MAX_CONCURRENT_STREAMS,
	ProtocolError,
	RECOMMENDED_CHUNK,
	RESET_INTERNAL_ERROR,
	buildClose,
	buildData,
	buildOpen,
	buildReset,
	buildWindow,
	encodeFrame,
	parseResetReason,
	parseWindowCredit,
} from "./framing";

export interface StreamHandle {
	id: number;
	write(data: Uint8Array): Promise<void>;
	close(): Promise<void>;
	reset(reason?: number): Promise<void>;
	/** Yields bytes as the peer sends DATA frames; ends on CLOSE/RESET. */
	read(): AsyncGenerator<Uint8Array, void, void>;
	/** Convenience: read until EOF into a single buffer. */
	readAll(): Promise<Uint8Array>;
	/** Resolved when the peer sends CLOSE or RESET, or the mux is torn down. */
	closed: Promise<void>;
}

type FrameSink = (bytes: Uint8Array) => void | Promise<void>;

interface StreamState {
	id: number;
	buffered: Uint8Array[];
	waiters: Array<(v: Uint8Array | null) => void>;
	send_credit: number;
	recv_credit: number;
	unacked_recv: number;
	writer_closed: boolean;
	reader_closed: boolean;
	creditWaiters: Array<() => void>;
	closedResolvers: Array<() => void>;
}

export class Multiplexer {
	private decoder = new FrameDecoder();
	private streams = new Map<number, StreamState>();
	private nextLocalId: number;
	private closed = false;

	constructor(
		private send: FrameSink,
		opts?: { isListener?: boolean },
	) {
		this.nextLocalId = opts?.isListener ? 2 : 1;
	}

	/** Feed plaintext bytes from the TLS layer into the framing decoder. */
	feed(plaintext: Uint8Array): void {
		if (!plaintext.byteLength) return;
		this.decoder.feed(plaintext);
		while (true) {
			let frame: ReturnType<FrameDecoder["next"]>;
			try {
				frame = this.decoder.next();
			} catch (err) {
				if (err instanceof ProtocolError) {
					this.tearDown();
					return;
				}
				throw err;
			}
			if (!frame) return;
			this.dispatch(frame);
		}
	}

	async openStream(initial: Uint8Array = new Uint8Array(0)): Promise<StreamHandle> {
		if (this.closed) throw new Error("mux closed");
		if (this.streams.size >= MAX_CONCURRENT_STREAMS) {
			throw new Error("concurrent stream cap reached");
		}
		const id = this.allocLocalId();
		const state: StreamState = {
			id,
			buffered: [],
			waiters: [],
			send_credit: INITIAL_WINDOW,
			recv_credit: INITIAL_WINDOW,
			unacked_recv: 0,
			writer_closed: false,
			reader_closed: false,
			creditWaiters: [],
			closedResolvers: [],
		};
		if (initial.byteLength) {
			state.send_credit -= initial.byteLength;
		}
		this.streams.set(id, state);
		await this.emit(buildOpen(id, initial));
		return this.makeHandle(state);
	}

	tearDown(): void {
		if (this.closed) return;
		this.closed = true;
		for (const state of this.streams.values()) {
			for (const w of state.waiters) w(null);
			state.waiters = [];
			for (const r of state.closedResolvers) r();
			state.closedResolvers = [];
			for (const w of state.creditWaiters) w();
			state.creditWaiters = [];
		}
		this.streams.clear();
	}

	// --- internals ---

	private allocLocalId(): number {
		const id = this.nextLocalId;
		this.nextLocalId += 2;
		return id;
	}

	private makeHandle(state: StreamState): StreamHandle {
		const mux = this;
		return {
			id: state.id,
			async write(data: Uint8Array) {
				if (state.writer_closed) throw new Error(`stream ${state.id} closed`);
				let view = data;
				while (view.byteLength > 0) {
					const chunkLen = Math.min(view.byteLength, RECOMMENDED_CHUNK, state.send_credit);
					if (chunkLen <= 0) {
						await new Promise<void>((resolve) => state.creditWaiters.push(resolve));
						continue;
					}
					const chunk = view.slice(0, chunkLen);
					view = view.slice(chunkLen);
					state.send_credit -= chunkLen;
					await mux.emit(buildData(state.id, chunk));
				}
			},
			async close() {
				if (state.writer_closed) return;
				state.writer_closed = true;
				await mux.emit(buildClose(state.id));
			},
			async reset(reason = RESET_INTERNAL_ERROR) {
				if (state.writer_closed && state.reader_closed) return;
				state.writer_closed = true;
				state.reader_closed = true;
				await mux.emit(buildReset(state.id, reason));
				mux.streams.delete(state.id);
				for (const r of state.closedResolvers) r();
			},
			async *read() {
				while (true) {
					if (state.buffered.length > 0) {
						yield state.buffered.shift() as Uint8Array;
						continue;
					}
					if (state.reader_closed) return;
					const next = await new Promise<Uint8Array | null>((resolve) =>
						state.waiters.push(resolve),
					);
					if (next === null) return;
					yield next;
				}
			},
			async readAll() {
				const parts: Uint8Array[] = [];
				for await (const chunk of this.read()) parts.push(chunk);
				let total = 0;
				for (const p of parts) total += p.byteLength;
				const out = new Uint8Array(total);
				let off = 0;
				for (const p of parts) {
					out.set(p, off);
					off += p.byteLength;
				}
				return out;
			},
			closed: new Promise<void>((resolve) => state.closedResolvers.push(resolve)),
		};
	}

	private dispatch(frame: { stream_id: number; flags: number; payload: Uint8Array }): void {
		if (frame.flags & FLAG_OPEN) {
			// As the dialer we never expect peer OPENs in v1. Spec-conformant
			// behavior is to RESET any stream the peer tries to open.
			void this.emit(buildReset(frame.stream_id, 0x01));
			return;
		}
		const state = this.streams.get(frame.stream_id);
		if (!state) {
			void this.emit(buildReset(frame.stream_id, 0x01));
			return;
		}
		if (frame.flags & FLAG_DATA) {
			if (frame.payload.byteLength > state.recv_credit) {
				void this.emit(buildReset(frame.stream_id, 0x02));
				this.terminate(state);
				return;
			}
			state.recv_credit -= frame.payload.byteLength;
			state.unacked_recv += frame.payload.byteLength;
			if (state.waiters.length > 0) {
				const w = state.waiters.shift() as (v: Uint8Array | null) => void;
				w(frame.payload);
			} else {
				state.buffered.push(frame.payload);
			}
			if (state.unacked_recv >= INITIAL_WINDOW / 2) {
				const grant = state.unacked_recv;
				state.recv_credit += grant;
				state.unacked_recv = 0;
				void this.emit(buildWindow(frame.stream_id, grant));
			}
		}
		if (frame.flags & FLAG_CLOSE) {
			state.reader_closed = true;
			while (state.waiters.length) {
				const w = state.waiters.shift() as (v: Uint8Array | null) => void;
				w(null);
			}
			if (state.writer_closed) this.terminate(state);
			for (const r of state.closedResolvers) r();
			state.closedResolvers = [];
		}
		if (frame.flags & FLAG_WINDOW) {
			try {
				const credit = parseWindowCredit(frame);
				state.send_credit += credit;
			} catch {
				void this.emit(buildReset(frame.stream_id, 0x01));
				this.terminate(state);
				return;
			}
			while (state.creditWaiters.length > 0 && state.send_credit > 0) {
				const w = state.creditWaiters.shift() as () => void;
				w();
			}
		}
		if (frame.flags & FLAG_RESET) {
			try {
				parseResetReason(frame);
			} catch {
				// tolerate unknown reasons per spec
			}
			state.reader_closed = true;
			while (state.waiters.length) {
				const w = state.waiters.shift() as (v: Uint8Array | null) => void;
				w(null);
			}
			this.terminate(state);
		}
	}

	private terminate(state: StreamState): void {
		for (const r of state.closedResolvers) r();
		state.closedResolvers = [];
		for (const w of state.creditWaiters) w();
		state.creditWaiters = [];
		this.streams.delete(state.id);
	}

	private async emit(frame: {
		stream_id: number;
		flags: number;
		payload: Uint8Array;
	}): Promise<void> {
		if (this.closed) return;
		await this.send(encodeFrame(frame));
	}
}
